pkg/test/test.go

```diff
@@ -2,98 +2,121 @@ package test
-	if len(configFilenames) == 0 {
-		configFilenames = append(configFilenames, cfgFilename)
-	}
+func newMocks() *Mocks {
+	components := make(map[string]mockComponent)
+	dockerRunner := newDockerRunner()
 
-	for _, filename := range configFilenames {
-		log.Println(fmt.Sprintf("booting configuration %s", filename))
-		bootFromFile(filename)
+	return &Mocks{
+		components:   components,
+		dockerRunner: dockerRunner,
 	}
+}
+
+func (t *Mocks) Boot(config cfg.Config) {
+	t.bootFromConfig(config)
 
-	wait.Wait()
+	t.waitGroup.Wait()
 
 	log.Println("test environment up and running")
-	fmt.Println()
+	log.Println()
 }
 
-func bootFromFile(filename string) {
-	config := readConfig(filename)
+func (t *Mocks) bootFromConfig(config cfg.Config) {
+	mocks := config.GetStringMap("mocks")
 
-	for _, mockConfig := range config.Mocks {
-		bootComponent(mockConfig)
+	for name, _ := range mocks {
+		settings := &mockSettings{}
+		key := fmt.Sprintf("mocks.%s", name)
+		config.UnmarshalKey(key, settings)
+
+		component := t.createComponent(settings.Component)
+
+		t.components[name] = component
+		component.Boot(name, config, settings)
+		t.runComponent(component)
 	}
 }
 
-func bootComponent(mockConfig configInput) {
-	component := mockConfig["component"]
-	name := mockConfig["name"].(string)
-
+func (t *Mocks) createComponent(component string) mockComponent {
 	switch component {
+	case "mysql":
+		return &mysqlComponent{}
+	case "sns_sqs":
+		return &snsSqsComponent{}
 	case "cloudwatch":
-		runCloudwatch(name, mockConfig)
+		return &cloudwatchComponent{}
 	case "dynamodb":
-		runDynamoDb(name, mockConfig)
+		return &dynamoDbComponent{}
 	case "elasticsearch":
-		runElasticsearch(name, mockConfig)
+		return &elasticsearchComponent{}
 	case "kinesis":
-		runKinesis(name, mockConfig)
-	case "mysql":
-		runMysql(name, mockConfig)
-	case "redis":
-		runRedis(name, mockConfig)
-	case "sns_sqs":
-		runSnsSqs(name, mockConfig)
+		return &kinesisComponent{}
 	case "wiremock":
-		runWiremock(name, mockConfig)
+		return &wiremockComponent{}
+	case "redis":
+		return &redisComponent{}
 	default:
-		err := fmt.Errorf("unknown component '%s'", component)
-		logErr(err, err.Error())
+		panic(fmt.Errorf("unknown component type: %s", component))
 	}
 }
 
-func Shutdown() {
-	for _, res := range dockerResources {
-		if err := dockerPool.Purge(res); err != nil {
-			log.Fatalf("Could not purge resource: %s", err)
-		}
+func (t *Mocks) runComponent(component mockComponent) {
+	t.waitGroup.Add(1)
+	go func() {
+		defer t.waitGroup.Done()
+		component.Run(t.dockerRunner)
+	}()
+}
+
+func (t *Mocks) ProvideClient(name string, clientType string) interface{} {
+	component := t.components[name]
+	return component.ProvideClient(clientType)
+}
+
+func (t *Mocks) Shutdown() {
+	t.dockerRunner.PurgeAllResources()
+}
+
+func Boot(configFilenames ...string) *Mocks {
+	if len(configFilenames) == 0 {
+		configFilenames = append(configFilenames, "config.test.yml")
```

Comment:
should we rather prepend config.test.yml? Then a user of the component can still decide to override the debug flag for a component for example

Reply:
The code here is misleading.
We are not appending anything, we are using a default singleton slice.
```go
	if len(configFilenames) == 0 {
		configFilenames = []string{"config.test.yml"}
	}
```

Reply:
✅